File: usr/projects/quissme/quissme-backend/app/core/buff_system.py
"""Buff system for gamification"""
from typing import Dict, List, Any
from datetime import datetime, timedelta, timezone
import logging

logger = logging.getLogger(__name__)

# Buff definitions
BUFF_DEFINITIONS = {
    "liebesflüsterer": {
        "name": "Liebesflüsterer",
        "duration_days": 7,
        "effect": "+15% Empathie",
        "description": "Erhöhte Fähigkeit, deinen Partner zu verstehen und einfühlsam zu kommunizieren.",
        "icon": "💕",
    },
    "harmonie_welle": {
        "name": "Harmonie-Welle",
        "duration_days": 3,
        "effect": "-Konflikte, +Geduld",
        "description": "Reduzierte Konflikte und erhöhte Geduld in schwierigen Momenten.",
        "icon": "🌊",
    },
    "versöhnungs_kraft": {
        "name": "Versöhnungs-Kraft",
        "duration_days": 1,
        "effect": "Nach Streit: schneller Reset",
        "description": "Hilft euch, schneller nach Konflikten wieder zusammenzufinden.",
        "icon": "🕊️",
    },
    "neugier_funkeln": {
        "name": "Neugier-Funkeln",
        "duration_days": 7,
        "effect": "+Fragen stellen, -Assumieren",
        "description": "Mehr Interesse füreinander und weniger Annahmen.",
        "icon": "✨",
    },
    "intimitäts_boost": {
        "name": "Intimitäts-Boost",
        "duration_days": 7,
        "effect": "+Körperliche Nähe",
        "description": "Erhöhte Lust auf körperliche Nähe und Zärtlichkeit.",
        "icon": "🔥",
    },
    "wert_schärfer": {
        "name": "Wert-Schärfer",
        "duration_days": 14,
        "effect": "Klarere Prioritäten",
        "description": "Besseres Verständnis für gemeinsame Werte und Prioritäten.",
        "icon": "💎",
    },
    "vertrauen_schild": {
        "name": "Vertrauen-Schild",
        "duration_days": 7,
        "effect": "+Sicherheit",
        "description": "Erhöhtes Vertrauen und Sicherheit in der Beziehung.",
        "icon": "🛡️",
    },
    "abenteuer_ruf": {
        "name": "Abenteuer-Ruf",
        "duration_days": 7,
        "effect": "+Neue Erfahrungen",
        "description": "Lust auf neue gemeinsame Abenteuer und Erfahrungen.",
        "icon": "🗺️",
    },
}


def get_buff_definition(buff_type: str) -> Dict[str, Any]:
    """
    Get buff definition.
    
    Args:
        buff_type: Buff type identifier
        
    Returns:
        Buff definition
        
    Raises:
        ValueError: If buff type not found
    """
    if buff_type not in BUFF_DEFINITIONS:
        raise ValueError(f"Buff type {buff_type} not found")
    
    return BUFF_DEFINITIONS[buff_type]


def calculate_buff_expiration(buff_type: str) -> datetime:
    """
    Calculate buff expiration time.
    
    Args:
        buff_type: Buff type identifier
        
    Returns:
        Expiration datetime
    """
    buff_def = get_buff_definition(buff_type)
    duration_days = buff_def["duration_days"]
    
    return datetime.now(timezone.utc) + timedelta(days=duration_days)


def is_buff_expired(expires_at: datetime) -> bool:
    """
    Check if buff has expired.
    
    Args:
        expires_at: Expiration datetime
        
    Returns:
        True if expired, False otherwise
    """
    return datetime.now(timezone.utc) > expires_at


def get_buff_time_remaining(expires_at: datetime) -> str:
    """
    Get human-readable time remaining for buff.
    
    Args:
        expires_at: Expiration datetime
        
    Returns:
        Time remaining string
    """
    now = datetime.now(timezone.utc)
    
    if now > expires_at:
        return "Expired"
    
    remaining = expires_at - now
    days = remaining.days
    hours = remaining.seconds // 3600
    minutes = (remaining.seconds % 3600) // 60
    
    if days > 0:
        return f"{days}d {hours}h"
    elif hours > 0:
        return f"{hours}h {minutes}m"
    else:
        return f"{minutes}m"


def list_all_buffs() -> List[Dict[str, Any]]:
    """
    List all available buffs.
    
    Returns:
        List of buff definitions
    """
    return list(BUFF_DEFINITIONS.values())

File: usr/projects/quissme/quissme-frontend/src/system/BuffSystem.ts
// system/BuffSystem.ts - Buff State Management

import { create } from 'zustand';

export type BuffType = 
  | 'Liebesflüsterer'
  | 'Harmonie-Welle'
  | 'Versöhnungs-Kraft'
  | 'Neugier-Funkeln'
  | 'Intimitäts-Boost'
  | 'Wert-Schärfer';

export interface Buff {
  id: string;
  type: BuffType;
  description: string;
  effect: string;
  durationDays: number;
  startDate: Date;
  expiryDate: Date;
  icon: string;
  percentageBoost: number;
}

export interface BuffState {
  activeBuffs: Buff[];
  addBuff: (buff: Buff) => void;
  removeBuff: (buffId: string) => void;
  getActiveBuffs: () => Buff[];
  hasBuffType: (type: BuffType) => boolean;
  getBuffEffect: (type: BuffType) => number;
}

const BUFF_DEFINITIONS: Record<BuffType, Omit<Buff, 'id' | 'startDate' | 'expiryDate'>> = {
  'Liebesflüsterer': {
    type: 'Liebesflüsterer',
    description: 'Enhanced empathy and listening skills',
    effect: '+15% Empathie',
    durationDays: 7,
    icon: '💬',
    percentageBoost: 15
  },
  'Harmonie-Welle': {
    type: 'Harmonie-Welle',
    description: 'Reduced conflicts, increased patience',
    effect: '-Konflikte, +Geduld',
    durationDays: 3,
    icon: '🌊',
    percentageBoost: 20
  },
  'Versöhnungs-Kraft': {
    type: 'Versöhnungs-Kraft',
    description: 'Quick recovery after conflicts',
    effect: 'Schneller Reset nach Streit',
    durationDays: 1,
    icon: '🕊️',
    percentageBoost: 25
  },
  'Neugier-Funkeln': {
    type: 'Neugier-Funkeln',
    description: 'Better question asking, less assumptions',
    effect: '+Fragen stellen, -Assumieren',
    durationDays: 7,
    icon: '✨',
    percentageBoost: 18
  },
  'Intimitäts-Boost': {
    type: 'Intimitäts-Boost',
    description: 'Enhanced physical and emotional closeness',
    effect: '+Körperliche Nähe',
    durationDays: 7,
    icon: '💕',
    percentageBoost: 22
  },
  'Wert-Schärfer': {
    type: 'Wert-Schärfer',
    description: 'Clearer priorities and values alignment',
    effect: 'Klarere Prioritäten',
    durationDays: 14,
    icon: '🎯',
    percentageBoost: 20
  }
};

export const useBuffStore = create<BuffState>((set, get) => ({
  activeBuffs: [],

  addBuff: (buff: Buff) => {
    set((state) => ({
      activeBuffs: [...state.activeBuffs, buff]
    }));
  },

  removeBuff: (buffId: string) => {
    set((state) => ({
      activeBuffs: state.activeBuffs.filter(b => b.id !== buffId)
    }));
  },

  getActiveBuffs: () => {
    const state = get();
    const now = new Date();
    return state.activeBuffs.filter(buff => buff.expiryDate > now);
  },

  hasBuffType: (type: BuffType) => {
    const activeBuffs = get().getActiveBuffs();
    return activeBuffs.some(buff => buff.type === type);
  },

  getBuffEffect: (type: BuffType) => {
    const activeBuffs = get().getActiveBuffs();
    const buff = activeBuffs.find(b => b.type === type);
    return buff ? buff.percentageBoost : 0;
  }
}));

export function createBuff(type: BuffType): Buff {
  const definition = BUFF_DEFINITIONS[type];
  const startDate = new Date();
  const expiryDate = new Date(startDate.getTime() + definition.durationDays * 24 * 60 * 60 * 1000);

  return {
    id: `buff_${Date.now()}_${Math.random()}`,
    ...definition,
    startDate,
    expiryDate
  };
}
